Add reduced basin functionality

Wegatriespython opened the pull request:
### Reduced basin filtering for local nexus development

Adds basin filtering to enable running the nexus module locally for development.
Reduced basin mode is not intended for production, basin reduction is not
physically realistic. However for development of scenarios, model capabilities or policy design, the reduced basin scenario needs to give a good proxy of the whole model, hence the granular control over its implementation.

Addresses #414. Builds on #405 (now merged).

## Design

Basin selection is two-step and compositional:

1. Automatic selection per region: either `first_k` (head N from CSV order)
   or `stress` (N basins sampled across the demand/supply ratio spectrum).
2. filter_list augmentation: if specific basins are named, they are added
   to the automatic set (union, not replacement). E.g. automatic gives
   {A1, B1, C1}, filter_list [B1, B2, B3] gives {A1, B1, B2, B3, C1}.

All regions are guaranteed coverage by the automatic step.

## New CLI options (on `mix-models water-ix nexus`)

```
--reduced-basin / --no-reduced-basin
                                Toggle for A/B testing between full and
                                reduced basin sets.
--num-basins INTEGER            Basins per region (default 3).
--basin-selection [first_k|stress]
                                Automatic selection method (default first_k).
                                stress ranks basins by demand/supply ratio.
--filter-list TEXT              Additional basins on top of automatic
                                selection (repeatable).
```

### Python API

```python
# Toggle for A/B comment one line to switch
ctx.reduced_basin = True

ctx.basin_selection = "stress"   # or "first_k"
ctx.num_basins = 3               # per region
# Additive on automatic set:
ctx.filter_list = ["10|SAS", "115|SAS"]
```

## Files changed

| File | What changed |
|---|---|
| `utils.py` | `filter_basins_by_region()` (two-step selection), `compute_basin_demand_ratio()`, `_select_by_stress()`, `_diversity_select()` |
| `cli.py` | New click options on `nexus_cli` |
| `build.py` | Applies `filter_basins_by_region()` in `map_basin()`, stores `context.valid_basins` |
| `demands.py`, `infrastructure.py`, `irrigation.py`, `water_supply.py` | Filter data to `context.valid_basins` |
| `conftest.py` | `setup_valid_basins()` helper for tests |
| `test_utils.py` | Tests for stress selection, demand ratio, filter_list additive behavior |

## How to review

- Core logic: `filter_basins_by_region()` in `utils.py` - verify the two-step
  selection and the additive union of filter_list.
- Stress selection: `compute_basin_demand_ratio()` and `_select_by_stress()` -
  verify demand/supply ranking and diversity sampling.
- Data modules: check that `context.valid_basins` filtering is applied
  consistently across demands, infrastructure, irrigation, water_supply.
- CLI: verify options propagate correctly to context in `cli.py`.

## PR checklist

- [ ] Continuous integration checks all pass
- [x] Add or expand tests; coverage checks pass
- [x] Update doc/whatsnew

Wegatriespython committed:
Add basin filtering

Use basin filtering everywhere

More filtering

More filtering

Fix accidental removal

Add basin reduction options

Clean up filtering code

Add min 1 basin region pair req

Add reg_to_basin return tech

consolidate return logic

Fix failing tests

Tests failure due to missing valid_basins list. Adding list to context
to fix issue.

Wegatriespython committed:
Fix basin filtering for cooling mode

Wegatriespython committed:
Fix dtype for pandas

Wegatriespython committed:
Add demand/supply stress-based basin selection

Wegatriespython committed:
Fix lint

Wegatriespython committed:
Make filter_list additive on automatic basin selection

filter_list now augments the automatic selection (first_k or stress)
rather than replacing it. Docstring updated to document the two-step
design. Test parametrized over both selection modes.

adrivinca reviewed (changes requested):
I think this is a useful feature, both for quickly testing new feature on the water module without having extremely long runs, and for application that focus on specific regions (e.g. Eurpoean projects), where we can neglect the energy-water dynamics in other continents and reduce run time.

To me the proposed changes look good. I would ask to please update the doc/water files, in particular `index.rst`, to update the cli command options and, if possible check and update obsolete text and reference to utility functions.

Wegatriespython committed:
Fix pandas 3.0 compat in water data functions

- utils.py: groupby().apply() drops groupby column in pandas 3.0;
  replace with cumcount() boolean mask
- demands.py: cast StringDtype columns to object before xr.Dataset()
- water_supply.py: filter delineation and e-flow CSVs to valid_basins
  in add_e_flow (mirrors read_water_availability pattern)

Wegatriespython committed:
Refactor water tests: add water_params factory, reduced-basin coverage

Centralize region→type_reg mapping in water_params() factory. Replace
all inline param dicts. Add reduced_basin=True variants across data
tests. Refactor test_irrigation to use water_context fixture.

Wegatriespython committed:
Update docs and whatsnew for basin filtering

Wegatriespython committed:
Add FIXME for indirect region derivation in water_for_ppl

adrivinca reviewed (approved):
To me it looks good, also thanks for the documentation udates.
Just nee to clarify why tests fail, maybe rebase needed?

khaeru commented:
> Just nee to clarify why tests fail, maybe rebase needed?

@adrivinca @Wegatriespython —this failure is new and not related to this PR. See https://github.com/iiasa/message-ix-models/issues/469#issuecomment-4011332645 and #482, just opened, which should resolve it.

After I merge that one commit to `main`, this branch can be rebased and the failure should disappear.

khaeru reviewed (approved):
I have not looked in detail at the code, but appreciate the expanded tests covering this new functionality and I see that comments from @adrivinca have been addressed.

Two comments that do not block merging:
1. I found the text/examples (under "Design" and "Python API") in the PR description to be helpful. In the future it would be great to simply put such examples in the documentation directly. That way, people finding and reading the docs do not need to hunt back for older PRs to get a complete description of how the code is supposed to work.
2. The "Python API" example seems to indicate there is a growing collection of Context settings that are recognized by `.model.water`. It may make sense to add a class like `.model.water.config.Config`, analogous to `.model.transport.config.Config` or `.report.config.Config`, that collects all settings that are internal to this module. Then one instance of this class can be stored/expected at `Context.water`.

This will be good to merge once CI checks all pass.
